universal-application-tool-0.0.1/app/controllers/applicant/ApplicantProgramReviewController.java

```diff
@@ -0,0 +1,129 @@
+    this.summaryView = checkNotNull(summaryView);
+    this.profileUtils = checkNotNull(profileUtils);
+  }
+
+  @Secure
+  public CompletionStage<Result> review(Request request, long applicantId, long programId) {
+    return checkApplicantAuthorization(profileUtils, request, applicantId)
+        .thenComposeAsync(
+            v -> applicantService.getReadOnlyApplicantProgramService(applicantId, programId),
+            httpExecutionContext.current())
+        .thenApplyAsync(
+            (roApplicantProgramService) -> {
+              ImmutableList<SummaryData> summaryData = roApplicantProgramService.getSummaryData();
+              // TODO: Get program title. (Currently no way to do that from
+              // roApplicantProgramService)
+              String programTitle = "Program title";
+              return (summaryData.size() > 0)
+                  ? ok(
+                      summaryView.render(
+                          request, applicantId, programId, programTitle, summaryData))
+                  : notFound();
+            },
+            httpExecutionContext.current())
+        .exceptionally(
+            ex -> {
+              if (ex instanceof CompletionException) {
+                Throwable cause = ex.getCause();
+                if (cause instanceof SecurityException) {
+                  return unauthorized();
+                }
+                if (cause instanceof ProgramNotFoundException) {
+                  return notFound(cause.toString());
+                }
+                throw new RuntimeException(cause);
+              }
+              throw new RuntimeException(ex);
+            });
+  }
+
+  @Secure
+  public CompletionStage<Result> submit(Request request, long applicantId, long programId) {
+    return checkApplicantAuthorization(profileUtils, request, applicantId)
+        .thenComposeAsync(
+            v -> {
+              return submit(applicantId, programId);
+            },
+            httpExecutionContext.current())
+        .exceptionally(
+            ex -> {
+              if (ex instanceof CompletionException) {
+                Throwable cause = ex.getCause();
+                if (cause instanceof SecurityException) {
+                  return unauthorized();
+                }
+                if (cause instanceof ProgramNotFoundException) {
+                  return notFound(cause.toString());
+                }
+                throw new RuntimeException(cause);
+              }
+              throw new RuntimeException(ex);
+            });
+  }
+
+  private CompletionStage<Result> submit(long applicantId, long programId) {
+    Call endOfProgramSubmission = routes.ApplicantProgramsController.index(applicantId);
+    return applicationRepository
+        .submitApplication(applicantId, programId)
+        .thenApplyAsync(
+            applicationMaybe -> {
+              if (applicationMaybe.isEmpty()) {
+                return found(endOfProgramSubmission)
```

Comment:
suggestion (blocking): If there was an error saving the application, I'd expect to stay on the review page so that I can try again - is that feasible here?